### ISSUE TITLE
Can't get CookieLife MaxAge to work

### ISSUE THREAD

**Issue by lefthandedgoat** [SuaveIO/suave#338]
I am trying to use an expiring cookie instead of a Session cookie.  I can't find any tests for it, and I cant debug the test on OSX or Win10 with VS 2015.  It seems to hang when starting.

My relevant code
```
let twoWeeks = System.TimeSpan.FromDays(14.0)
let twoWeekCookie = Suave.Cookie.CookieLife.MaxAge twoWeeks
```
After someone successfully logs in I have this code to set the cookie
```
Auth.authenticated twoWeekCookie false
>>= stateful twoWeekCookie false
>>= sessionStore (fun store -> store.set "user_id" id)
 >>= request (fun _ -> FOUND <| paths.home_link validUser.Name)
```

It blows up on this line: https://github.com/SuaveIO/suave/blob/master/src/Suave/Cookie.fs#L36

I pulled some of the cookie source into my project and I was able to print what it was trying to parse:

```
auth=WslRJY!CgISXmc4XNtE!lJCCiCy9PepAm0xCBQ4HosYizWK!mxsP!1fP!TEdmJyPvnPlJFJIxuPmNRAEX3Ab7ntV220dNlHr7Vf9BX2CgANz_rpuoT6_zWovflc8ndyXaxd9A2fs7lNjRyWy70QGZzSiSBQqRsIMQ1Z6OBNccWzwNhFN52NHuyObXkjES36pu8eNoRff0bWWrT3j4bnjHjVTg9OZOr5Qe6OV8RQzv8Xyjq1h2YobcoITzd44r2sr_4sfGP9sbfBcoRkfgL7x01PVSJaRDCUMQkxtQqK82_M$;Path=/;Expires=Sat
```
Relevent bit at the end 
```
;Path=/;Expires=Sat
```
Looks like the `Expires` part of the cookie is just being set with the first part of the date, and not the full date which would have been something like `Sat, 27 Nov 2015 05:00:00 GMT`

Hope this helps get you pointed in the right direction.

**Comment by lefthandedgoat** [comment]
I noticed this issue `Tests are just silent if port is taken #329` and that was the reason I could not run the tests.  I have them running now and I will spend some time today trying to figure things out.

**Comment by lefthandedgoat** [comment]
I found the issue here, its splitting on comma and that is splitting out the data `Sat, 27 Nov...` and breaking up the value of the Expires cookie:

https://github.com/SuaveIO/suave/blob/master/src/Suave/Cookie.fs#L72

Here is an `Auth.fs` test that will reproduce the issue:

```
    testCase "can set cookie" <| fun _ ->
      let timespan = System.TimeSpan.FromDays(13.0)
      let cookieLife = Suave.Cookie.CookieLife.MaxAge timespan
      let ctx = runWithConfig (Auth.authenticated cookieLife false >>= OK "ACK")
      let cookies = ctx |> reqCookies' HttpMethod.GET "/"  None
```

I do not know off the top of my head what the right solution is, without adding possible regressions.

My brain is tired from walking through all the code.  If I think of an appropriate solution I will let you know, but for now I hope you guys can =).

**Comment by lefthandedgoat** [comment]
I fixed up that line enough to make it work for me locally, and the cookie is now set with the right expiration.  My original problem that led me to try expiration based cookies still exists though:  My state cookie does not survive a stop/start of the server, forcing me to log back in after each code build =(

**Comment by ademar** [comment]
Cool, I was about to tell you about the server key :)

Can you send us a PR with the fix for the cookie issue?

Cheers.

**Comment by lefthandedgoat** [comment]
I would but it wouldn't work for the line was originally intended for, for splitting out duplicate headers.  I haven't thought of a solution that would work for both scenarios.

**Comment by ademar** [comment]
The fix is simply to remove that line; since the statement made in the comment is false; we do not concatenate equal headers.

In fact the only way of setting cookies with different `Expire` values is to send multiple "Set-Cookie" headers. Also according to RFC we should use `Max-Age` instead of  `Expire`.

for reference: 
https://github.com/scalaj/scalaj-http/issues/83
http://stackoverflow.com/a/6375214/48401

**Comment by haf** [comment]
The comment is not false. If do `Set-Cookie` twice, it may end up on the same header line.

**Comment by ademar** [comment]
@haf can you show where are we concatenating headers?

**Comment by haf** [comment]
https://github.com/SuaveIO/suave/blob/master/src/Suave/Cookie.fs#L117

**Comment by ademar** [comment]
I do not see any comma separated concatenation on that line.

On Sunday, 15 November 2015, Henrik Feldt <notifications@github.com> wrote:

> https://github.com/SuaveIO/suave/blob/master/src/Suave/Cookie.fs#L117
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SuaveIO/suave/issues/338#issuecomment-156816824>.
>


-- 
Ademar Gonzalez
+1-647-891-3606
http://ademar.name
https://github.com/ademar

**Comment by haf** [comment]
Well, it has that effect if you follow the code.

**Comment by ademar** [comment]
> Well, it has that effect if you follow the code.

Sorry, it does not. `Writers.addHeader` does not concatenate headers.

Also you are conflating request headers with response headers so no we shall not expect them as input in the response header collection.

Furthermore it is an obvious bug as shown by the top poster.

**Comment by haf** [comment]
**Responses**: I just checked up the writing of cookie headers. You're right that we are not comma-separating values, but a webserver may, according to [RFC2109 section 4.2.2](https://tools.ietf.org/html/rfc2109). However, that was changed in [RFC6265 section 3](https://tools.ietf.org/html/rfc6265) to disallow *folding* as they call it. So our way of writing responses is correct respect to RFC6265. I had read only the previous RFC when designing this.

Agree; remove the `collect` line.
